### DISCUSSION TITLE
Adapting Airwindows Galactic for use in Sapphire

### DISCUSSION THREAD

**Discussion by cosinekitty** [cosinekitty/sapphire#52] [General]
Hey @baconpaul and @airwindows, I thought I would start a general discussion about the topic of adapting Airwindows Galactic for use in future Sapphire modules. Currently, I am close to having a working prototype that is basically the same thing you two did for bringing Airwindows to VCV Rack, only with a different panel design bolted on top.

First, about naming it. I am provisionally calling it "Galaxy" for a couple of reasons. First, to honor Galactic by being similarly named, but not an identical name because I plan on experimenting with additional modulation controls (to be determined). However, @airwindows, please let me know if you would rather I call it something else as it starts to drift in its own direction. I have a backup name in mind: Sapphire Enormity. Asked from another direction, so long as I credit Airwindows Galactic, is it OK if I call it Enormity?

Omri Cohen has encouraged me to create modules where I start to integrate some of the chaotic oscillators I have created (Sapphire Glee/Frolic) as internal blocks to supply user-controlled but mostly automatic parameter movement. I am especially interested in synth pad generators, as I have mentioned in earlier conversation. Putting these together could be amazing!

I will credit any use of Galactic (or any other components derived from other authors) in these hypothetical synth pad modules.

I just want to be respectful and show gratitude for finally finding a great reverb that sounds perfect for my needs.

**Comment by airwindows** [comment]
I'm glad you found it useful :) use whatever name you enjoy, in some places (Reaper internal FX) I've got devs literally porting over the plugins en masse and changing ALL the names even when not changing the plugins (sometimes that dev does change things, though). I would say if people want to know how it's done, if you direct them to YOUR starting point (Galactic) that honors the MIT license, which doesn't really demand that you have to share all your own tweaks, just where you got the open source code you started with.

So from me, I would say 'use whatever name you like and have fun trying variations!'. Paul is a programmer like yourself and has worked wonders in bringing Galactic to places, but as long as it's the original Galactic we're talking about that one's from me. If you are using Paul's code that makes the original retro VST files more easily portable then you're also using Paul's work. If you're working from the original VST2 codebase (or AU codebase) that is directly from me, with my blessing :)

**Comment by cosinekitty** [comment]
OK, great! Quick question: if someone hooks up a cable to the left input port, but not the right input, do you guys send the left input into both left and right (normalling)? I think I need to do that in my module when there is no cable connected to the right input.

**Reply by baconpaul** [reply]
Yes Chris stuff is all stereo so left in my copies at in and averages at our in the airwindows rack module

**Reply by baconpaul** [reply]
(It’s actually a bit more complicated than that in the rack module because we support all the topologies to be a mixmaster polyphonic insert also but in the monophonic and obvious polyphonic case it does what I said)

**Reply by cosinekitty** [reply]
With carefully crafted signals, I can now tell that my port of Galactic is not correct. It doesn't have the stereo richness of the original. So I need to do a deep dive and figure out where the mistakes are.

**Reply by airwindows** [reply]
You know it's designed to take in mono signals as stereo, yes? There's a vibrato thing in there which can give you two different vibrato offsets for a centered mono signal.

Big difference between that, and just one side :)

So if your port only goes into left (when right is unplugged) there's the problem right there :) Paul said to use his in mono it 'copies at in' (copies to the unconnected input) and averages at out (combines the two outputs to mono if there's only one output). This is the right answer, I think.

At any rate, it looks like what you want to do is what you said you would: if there's not two inputs, send the single input into both inputs, and then probably listen to the stereo and it should be the same as a centered signal in stereo :)

**Reply by cosinekitty** [reply]
I finally found all the discrepancies between my code and the original Airwindows Galactic. There were a few things I didn't initialize correctly, and had some L/R mix-ups in formulas. The effect was that the left and right channels were essentially independent monaural reverbs. Now I have a unit test that feeds a CD-format (44.1 kHz stereo) recording of my voice through Airwindows Galactic and my own Sapphire Galaxy. The output files from both are WAV files about 3 MB in size. Both output WAV files are binary identical now!

If I just wanted to use Galactic as-is, I wouldn't have bothered doing all this. What I'm hoping for is to keep refactoring until I can express an entire class of reverb algorithms concisely, in which Galactic is expressible with a tiny amount of notation. I can imagine some Python program that you give it a little text file called galactic.txt, and it spits out C++ code that behaves the same way as Airwindows Galactic. Then I (and others) could tinker with alternative reverb algorithms. Who knows, maybe it could spit out a block diagram too?

**Comment by cosinekitty** [comment]
I created a synth pad today using Galaxy. [Here is the video](https://www.youtube.com/watch?v=F4x5HxTmLXs). I love this reverb algorithm!

**Reply by baconpaul** [reply]
That sounds great
